### ISSUE TITLE
Use higher resolution timestamps for output dirs

### ISSUE THREAD

**Issue by bhoeckendorf** [facebookresearch/hydra#1607]
# 🚀 Feature Request

I'd like to suggest using higher resolution timestamps for automatic output folder names by default. Additionally, if a folder already exists, automatically append a running index to ensure unique folders.

## Motivation

When starting multiple remote HPC jobs, some of them by accident start within the same second. They then proceed to overwrite each other's results.

## Pitch

Easy to implement, helps a lot.

**Comment by jieru-hu** [comment]
Thanks for the feature request. This can be easily configured.
the [default config](https://github.com/facebookresearch/hydra/blob/master/hydra/conf/hydra/output/default.yaml) is
```
run:
  dir: outputs/${now:%Y-%m-%d}/${now:%H-%M-%S}
```

To get more precise time you can do somethins like
```
hydra:
  run:
    dir: outputs/${now:%Y-%m-%dT%H:%M:%S.%fZ}
```
the output dir you will get
```
$ tree outputs/
outputs/
└── 2021-05-05T15:11:28.867703Z
    └── my_app.log
```

check out our doc for more info (the following is 1.0, you can also check 1.1 doc if you are on Hydra 1.1)
https://hydra.cc/docs/configure_hydra/intro/#hydra-resolvers
https://hydra.cc/docs/configure_hydra/workdir

**Comment by bhoeckendorf** [comment]
Thanks, this is very helpful. I suspected that there would be an easy way to configure this. Would it make sense to make your config suggestion the default, thus potentially saving others a lot of trouble?

**Comment by omry** [comment]
Thanks for the suggestion.
While it will side-step an issue hitting a very small percentage of the users completely, it will force near 100% of the users to work with much longer directories by default.

I think the tradeoff is not worth it.

**Comment by bhoeckendorf** [comment]
Ok. I'll make sure to update my config according to what @jieru-hu helpfully suggested.

**Comment by eric-mitchell** [comment]
I'd like to re-open this issue, because using micro-second resolution timestamps doesn't solve the problem for me, and it seems like a band-aid to a more fundamental issue anyway. Enabling multiple runs to write to the same directory *by default* seems to make it very easy for users to shoot themselves in the foot (as evidenced by the original issue here).

For example, in my computing setup, microsecond-resolution timestamps still do not enforce unique output directories, because of jobs that get run and backgrounded. For example:

```
$ date "+%s %N" & date "+%s %N" & wait
[1] 10798
[2] 10799
1630987478 518511940
1630987478 518511914
```

Here, the two commands are run at the same time at the microsecond resolution (and in fact at tenth of a microsecond resolution as well). 

So, my question is: How can we **guarantee** that hydra runs get unique output directories?

**Comment by omry** [comment]
How about using [hydra.job.id or hydra.job.num](https://hydra.cc/docs/configure_hydra/job#hydrajobid) as a part of your path?

**Comment by eric-mitchell** [comment]
I'm using the default launcher, and hydra.job.id doesn't seem to actually be populated. Maybe I'm using it wrong (see below)? Regardless, to me it seems like hydra should **require** that output directories are unique, rather than just make this possible if the right configuration is used. Is there a reason why someone would want to overwrite output directories by default? It seems high risk and low reward to me.


Using hydra.job.id in my configuration causes a crash when running `python hydra_test.py`:

```
omegaconf.errors.InterpolationToMissingValueError: MissingMandatoryValue while resolving interpolation: Missing mandatory value: hydra.job.id
    full_key: hydra.run.dir
    object_type=dict
```

`hydra_test.py`:
```
import hydra


@hydra.main(config_path="testconfig", config_name="config")
def run(cfg):
    import pdb; pdb.set_trace()


if __name__ == '__main__':
    run()
```

`testconfig/config.yaml`:
```
hydra:
  run:
    dir: ./outputs/${hydra.job.id}
```

**Comment by omry** [comment]
> I'm using the default launcher, and hydra.job.id doesn't seem to actually be populated. 

Ah, yes - those are only available in multirun. If you want to launch multiple jobs anyway why are you not using it? as an example - the Joblib Sweeper should enable you to run multiple jobs in parallel.

> to me it seems like hydra should require that output directories are unique, rather than just make this possible if the right configuration is used. Is there a reason why overwriting an output directory (or simultaneously writing to the same directory!) by default is the desired behavior? It seems high risk and low reward to me.

There are many ways to use Hydra.
Do you have a concrete feature request? saying "I want things to be this and that" without suggesting alternative or describing the desired behavior (and considering all existing users) is not productive.
As an example: do you want Hydra to fail if the directory already exists? What do you tell people that wants to rerun the job in the same directory (for resume purposes)?
Alternatively, do you want Hydra to dynamically generate a new directory (e.g. foo/bar_2)? What do you tell people that want to inspect the generated directories programmatically?

It also sounds like the issues you are facing are coming from the fact that you are not using Hydra properly. If you want to launch multiple jobs you should use multirun.

If you insist on having your sweeper outside of Hydra, it's your responsibility to ensure it works. for example - you could create your a bunch of unique directories and pass them to your Hydra apps (e.g. `hydra.run.dir=foo/bar/1312`), or just issue a small sleep between the jobs to ensure they are not starting at exactly the same time.

**Comment by eric-mitchell** [comment]
> If you want to launch multiple jobs anyway why are you not using it?

I'm using slurm for job management. However, within each slurm job, I want to invoke multiple runs of my script with different arguments in parallel to maximize GPU utilization. As far as I can tell, the submitit launcher plugin does not support this use case (though I'd be happy to be corrected here).

> saying "I want things to be this and that" without suggesting alternative or describing the desired behavior (and considering all existing users) is not productive

In my comment, I was wondering if there was a particular reason why the current behavior was the default- I apologize if I acted entitled or demanding. Hydra has been a really useful piece of software and I'm grateful for the work that's gone into it! 

To answer your questions:

> do you want Hydra to fail if the directory already exists?

Yes

> What do you tell people that wants to rerun the job in the same directory (for resume purposes)?

Since resuming jobs is probably the less common use case, would a flag to denote the intention to allow the (potentially destructive) behavior of writing to an existing directory (e.g. `hydra.job.allow_existing=True`) be sufficient here? Guild uses a similar strategy with the `--restart` flag (see [here](https://my.guild.ai/t/command-run/146#restart-an-operation)).

> Alternatively, do you want Hydra to dynamically generate a new directory (e.g. foo/bar_2)?

To me, this seems sensible and easy to enable in one's config if it is the desired behavior, e.g. `hydra.job.prevent_name_collisions=True`. As an alternative, if a job is run with the same output directory as an existing job, a new directory could be created inside the existing folder (`.hydra/reruns` or something) and reruns could be stored there.

> What do you tell people that want to inspect the generated directories programmatically?

If I have:

```
$ tree outputs/
outputs/
├── 2021_09_05
│   └── run.log
├── 2021_09_06
│   └── run.log
└── 2021_09_07
    └── run.log
```
and I run a new job whose output directory would be `2021_09_07` (i.e. a collision), we create `.hydra/reruns` for that directory and put the new run in `outputs/2021_09_07/.hydra/reruns/0`. If I'm traversing this programmatically, it's pretty easy to iterate over my runs normally and add a quick check (`os.path.exists(f"{run_dir}/.hydra/reruns")`) for the existence of `.hydra/reruns` for each (normal) run directory, traversing them (or selecting only the latest, etc) if desired.

> It also sounds like the issues you are facing are coming from the fact that you are not using Hydra properly

Is there a way to launch a multirun on slurm, with more than one run per slurm job? Otherwise I'm not sure Hydra supports my use case.

The easiest solution I've found so far has been using a custom OmegaConf resolver to append a randomly-generated uuid to the output directory, like what is described [here](https://stackoverflow.com/a/64765395/791018). This is a pretty concise workaround for others with this issue. Perhaps adding such a UUID resolver to hydra in addition to `${now}` and others would be the easiest way to enable the desired behavior?
